### PR TITLE
fixes and a possible canvas addition

### PR DESCRIPTION
**Pull request by s3bk**
What I am not sure about is `fill_layout` (see #202). I added it because it was needed to push layouts directly, but currently I am working on removing the dependency on text altogether. (And instead push the glyph outlines).

### CODE REVIEW

File: swf/src/shapes.rs [hunk 1]
```diff
@@ -93,12 +93,12 @@ impl Shape {
     }
 
     #[inline]
-    fn first(&self) -> LineSegment {
+    fn first(&self) -> &LineSegment {
```
Comment (pcwalton) [inline swf/src/shapes.rs]: Why this change? `LineSegment` fits in a single SIMD register; there's little point in returning references.

**Comment by pcwalton** [comment]
`fill_layout` is perfectly fine. I was thinking of adding it myself actually.

File: swf/src/shapes.rs (hunk 1 above)
Comment (s3bk) [reply swf/src/shapes.rs]: There was an error and I simply applied the suggested fix. No other reason.

Comment (s3bk) [reply swf/src/shapes.rs]: Dereferenced it now instead.

File: swf/src/shapes.rs [hunk 2]
```diff
@@ -94,12 +94,12 @@ impl Shape {
 
     #[inline]
     fn first(&self) -> LineSegment {
-        self.outline.first().unwrap()
+        *self.outline.first().unwrap()
```
Comment (pcwalton) [inline swf/src/shapes.rs]: Will this still compile?

Comment (s3bk) [reply swf/src/shapes.rs]: It compiles for me (`cargo build -p pathfinder_swf`)

Comment (s3bk) [reply swf/src/shapes.rs]: without:
```
error[E0308]: mismatched types
   --> swf/src/shapes.rs:102:9
    |
101 |     fn last(&self) -> LineSegment {
    |                       ----------- expected `shapes::LineSegment` because of return type
102 |         self.outline.last().unwrap()
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |         |
    |         expected struct `shapes::LineSegment`, found reference
    |         help: consider dereferencing the borrow: `*self.outline.last().unwrap()`
    |
    = note: expected type `shapes::LineSegment`
               found type `&shapes::LineSegment`
```

**Comment by pcwalton** [comment]
Hmm, we shouldn't return a reference from `Outline::last()` then, but let's fix that in a follow up.

**Comment by s3bk** [comment]
```
    pub(crate) outline: Vec<LineSegment>, // Could be Vec<(start, end)>
```
not much that can be done here …
